src/aws4.ts

```diff
@@ -0,0 +1,141 @@
+import * as crypto from 'node:crypto';
+
+export type Options = {
+  path: '/';
+  body: string;
+  host: string;
+  method: 'POST';
+  headers: {
+    'Content-Type': 'application/x-www-form-urlencoded';
+    'Content-Length': number;
+    'X-MongoDB-Server-Nonce': string;
+    'X-MongoDB-GS2-CB-Flag': 'n';
+  };
+  service: string;
+  region: string;
+  date?: Date;
+};
+
+export type AwsSessionCredentials = {
+  accessKeyId: string;
+  secretAccessKey: string;
+  sessionToken: string;
+};
+
+export type AwsLongtermCredentials = {
+  accessKeyId: string;
+  secretAccessKey: string;
+};
+
+export type SignedHeaders = {
+  headers: {
+    Authorization: string;
+    'X-Amz-Date': string;
+  };
+};
+
+export interface AWS4 {
+  /**
+   * Created these inline types to better assert future usage of this API
+   * @param options - options for request
+   * @param credentials - AWS credential details, sessionToken should be omitted entirely if its false-y
+   */
+  sign(
+    this: void,
+    options: Options,
+    credentials: AwsSessionCredentials | AwsLongtermCredentials | undefined
+  ): SignedHeaders;
+}
+
+const getHash = (str: string): string => {
+  return crypto.createHash('sha256').update(str, 'utf8').digest('hex');
+};
+const getHmacArray = (key: string | Uint8Array, str: string): Uint8Array => {
+  return crypto.createHmac('sha256', key).update(str, 'utf8').digest();
+};
+const getHmacString = (key: Uint8Array, str: string): string => {
+  return crypto.createHmac('sha256', key).update(str, 'utf8').digest('hex');
+};
+
+const getEnvCredentials = () => {
+  const env = process.env;
+  return {
+    accessKeyId: env.AWS_ACCESS_KEY_ID || env.AWS_ACCESS_KEY,
+    secretAccessKey: env.AWS_SECRET_ACCESS_KEY || env.AWS_SECRET_KEY,
+    sessionToken: env.AWS_SESSION_TOKEN
+  };
+};
+
+const convertHeaderValue = (value: string | number) => {
+  return value.toString().trim().replace(/\s+/g, ' ');
+};
+
+export function aws4Sign(
+  this: void,
+  options: Options,
+  credentials: AwsSessionCredentials | AwsLongtermCredentials | undefined
```

Comment:
We only ever call this with credentials already fetched - could we make this explicitly required?  

The benefit of that is that we then don't need logic to fetch credentials from the environment, if unset.  Right now, we rely on the AWS SDK to fetch these credentials for us and it would be nice not to add this logic back into the driver 🙂

Reply:
Making creds explicitly required.